presto-spi/src/main/java/com/facebook/presto/spi/block/StructBuilder.java

```diff
@@ -0,0 +1,214 @@
+
+public class StructBuilder
+{
+    private final BlockBuilderStatus builderStatus;
+    private final BlockBuilder builder;
+    private final List<Type> parameterTypes;
+    private final Iterator<Type> typeIterator;
+
+    public static StructBuilder arrayBuilder(Type elementType)
+    {
+        return new StructBuilder(asList(elementType));
+    }
+
+    public static StructBuilder mapyBuilder(Type keyType, Type valueType)
+    {
+        return new StructBuilder(asList(keyType, valueType));
+    }
+
+    public static StructBuilder rowBuilder(List<Type> parameterTypes)
+    {
+        return new StructBuilder(parameterTypes);
+    }
+
+    public static StructBuilder builder(int expectedSize, List<Type> parameterTypes)
+    {
+        return new StructBuilder(expectedSize, parameterTypes);
+    }
+
+    private StructBuilder(List<Type> parameterTypes)
+    {
+        this(1024, parameterTypes);
+    }
+
+    StructBuilder(int expectedSize, List<Type> parameterTypes)
+    {
+        this(new BlockBuilderStatus(expectedSize, expectedSize), parameterTypes);
+    }
+
+    private StructBuilder(BlockBuilderStatus builderStatus, List<Type> parameterTypes)
+    {
+        if (parameterTypes.isEmpty()) {
+            throw new IllegalArgumentException("parameterTypes are empty");
+        }
+
+        this.builderStatus = builderStatus;
+        this.builder = new VariableWidthBlockBuilder(builderStatus);
+        this.parameterTypes = unmodifiableList(parameterTypes);
+        this.typeIterator = new CycleIterator();
+    }
+
+    public static Block readBlock(SliceInput input)
+    {
+        return new VariableWidthBlockEncoding().readBlock(input);
+    }
+
+    public Slice build()
+    {
+        int size = builderStatus.getSizeInBytes() + 5 * (builder.getPositionCount() + 1);
+        DynamicSliceOutput output = new DynamicSliceOutput(size);
+        builder.getEncoding().writeBlock(output, builder.build());
+
+        return output.slice();
+    }
+
+    public StructBuilder addAll(Object... values)
+    {
+        for (Object value : values) {
+            add(value);
+        }
+
+        return this;
+    }
+
+    public StructBuilder addAll(List<?> values)
+    {
+        values.forEach(this::add);
+
+        return this;
+    }
+
+    public StructBuilder add(Block block, int position)
+    {
+        if (block.isNull(position)) {
+            builder.appendNull();
+        }
+        else {
+            block.writeBytesTo(position, 0, block.getLength(position), builder);
+            builder.closeEntry();
+        }
+        return this;
+    }
+
+    public StructBuilder addAll(Block block)
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            add(block, i);
+        }
+
+        return this;
+    }
+
+    public StructBuilder add(Object value)
+    {
+        if (!typeIterator.hasNext()) {
+            throw new IllegalArgumentException("Cannot add element on a unknown type");
```

Comment:
This should be `IllegalStateException` since the error depends on the state of the object and not on the argument